gt/main.c

```diff
@@ -244,13 +246,105 @@ print_ip_err_msg(struct gt_packet_headers *pkt_info)
 		"gt: receiving a packet with IP source address %s, and destination address %s, whose destination IP address is not the Grantor server itself.!\n",
 		src, dst);
 }
+static void
+gt_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	if (pcall_again == NULL) {
+		clear_ether_cache(eth_cache);
+		return;
+	}
+
+	/*
+	 * Deal with concurrency control by sequential lock
+	 * on the nexthop entry.
+	 */
+	write_seqlock(&eth_cache->lock);
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+	write_sequnlock(&eth_cache->lock);
+
+	*pcall_again = true;
+}
+
+/*
+ * Fill up the Ethernet cached header.
+ * Note that the destination MAC address should be filled up by LLS.
+ * The fields @ref_cnt and @ip_addr are not used by GT blocks.
```

Comment:
We need to use the `ref_cnt` because we don't know when the LLS block will stop writing to the Ethernet cache entry.

Also, GT will need to allocate the Ethernet cache entries and call `rte_add_key_data()` because it needs to make sure that the LLS block has released the entry before reusing it.